core/unix/signal.c

```diff
@@ -6631,11 +6633,120 @@ execute_native_handler(dcontext_t *dcontext, int sig, sigframe_rt_t *our_frame)
+     * for sig.
+     */
+    bool reuse_cur_frame = false;
+    if (dcontext == NULL && doing_detach) {
+        ASSERT(cur_xsp != NULL);
+        reuse_cur_frame = true;
+        if (!IS_RT_FOR_APP(info, sig)) {
+            /* If the app wants a non-RT frame, we cannot reuse the DR signal frame
+             * which is RT.
+             */
+            reuse_cur_frame = false;
+        } else if (APP_HAS_SIGSTACK(info)) {
+            bool currently_on_sigstack = cur_xsp != NULL &&
+                info->app_sigstack.ss_sp < (void *)cur_xsp &&
+                (void *)cur_xsp < info->app_sigstack.ss_sp + info->app_sigstack.ss_size;
+            if (!currently_on_sigstack) {
+                /* If the app has a sigstack, we must be on it. */
+                ASSERT_NOT_REACHED();
+                /* In release build, we bail on this optimization and try to deliver the
+                 * signal using a frame copy.
+                 */
+                reuse_cur_frame = false;
+            }
+            if (!USE_APP_SIGSTACK(info, sig)) {
+                /* If the app does not want to use the signal stack, we cannot
+                 * reuse the current frame.
+                 */
+                reuse_cur_frame = false;
+            }
+        }
+    }
 
+    /* Get the signals that are supposed to be blocked during handling of sig. */
+    kernel_sigset_t blocked;
     blocked = info->sighand->action[sig]->mask;
     if (!TEST(SA_NOMASK, (info->sighand->action[sig]->flags)))
         kernel_sigaddset(&blocked, sig);
+
+    RSTATS_INC(num_signals);
+    RSTATS_INC(num_native_signals);
+#ifdef X86_64
+    if (reuse_cur_frame) {
+        /* We've determined that we can reuse the current signal frame to
+         * deliver sig to the native signal handler. We want to simply jump
+         * to the handler, without a sigreturn. This way we preserve the app
+         * state currently in the frame (reg values, blocked sigmask).
+         */
+        /* Emulate the app's sigmask for this signal.
+         */
+        sigprocmask_syscall(SIG_SETMASK, &blocked, NULL, sizeof(blocked));
+
+        void (*asm_jmp_tgt)() = SIGACT_PRIMARY_HANDLER(&sigact_struct);
+        kernel_siginfo_t *siginfo_var = &our_frame->info;
+        kernel_ucontext_t *ucontext_var = &our_frame->uc;
+        asm volatile(
+            "mov %[jmp_tgt],  %%rcx\n\t"
+            "mov     %[sig],  %%rdi\n\t"
+            "mov %[siginfo],  %%rsi\n\t"
+            "mov %[ucontext], %%rdx\n\t"
+            "mov %[target_sp], %%rsp\n\t"
+            "jmp *%%rcx\n\t"
+            :
+            : [jmp_tgt] "m"(asm_jmp_tgt), [sig] "m"(sig), [siginfo] "m"(siginfo_var),
+              [ucontext] "m"(ucontext_var), [target_sp] "m"(cur_xsp)
+            /* Cannot add "rsp" as listing it in clobber list is deprecated. */
+            : "rdi", "rsi", "rdx", "rcx");
+        ASSERT_NOT_REACHED();
+    }
+#else
+    /* TODO i#6814: Add support for other archs. */
```

Comment:
I believe call_switch_stack() would work here if the handler took just one arg, and give us cross-arch implementations: or maybe it pushes a retaddr after switching. Hmm. Too bad using sigreturn requires a trampoline on the other side with some data passed through.

Will aarch64 be added ASAP?

Reply:
> Yes: call_switch_stack uses blx/blr/CALLC1 which uses call.

Though only the call is a concern (since blx/blr don't push to stack).

Reply:
Added the AArch64 version. Was easier to test it when I modified detach_signal to reproduce the issue.